### PR TITLE
Download models with pooch

### PR DESCRIPTION
**Pull request by GenevieveBuckley**
I've made a start on https://github.com/computational-cell-analytics/micro-sam/issues/36

* This PR downloads the models using pooch
* We are still using the SHA256 hashes for validation, although at some future point we could use a (non-cryptographic) XXH128 hash [as discussed here](https://github.com/computational-cell-analytics/micro-sam/issues/36#issuecomment-1809426270)
* I've removed the 'checkpoint' kwarg from `get_sam_model`. I think it makes more sense to have custom weights of any sort go through the `get_custom_sam_model` function (I haven't touched this function yet), and also if we download weights from different checkpoints then the hash validation is going to break.
* Problem: I can't test a lot of this, because...
    * we have low test coverage, and
    * the example scripts often use hard coded paths to locations on Constantin's computer, so I can't run them (eg: `examples/annotator_with_custom_model.py`)

### CODE REVIEW

**Comment by GenevieveBuckley** [comment]
If/when we go ahead with https://github.com/computational-cell-analytics/micro-sam/pull/276, we'll need to make sure `pooch.os_cache` is recomputed in the pooch MODELS registry every time, instead of just once when the module is imported.

File: micro_sam/util.py
Comment (constantinpape) [inline micro_sam/util.py:52]: Should we maybe wrap all of this into function(s)? 
This seems to be a good idea in general and is the easiest solution I can think of for updating the cache path re #277.

And we should probably have two functions: one that returns the cache path (and that can also be used by other functionality that uses it) and one that returns the pooch model registry.

**Comment by constantinpape** [comment]
Overall this is great!

> * We are still using the SHA256 hashes for validation, although at some future point we could use a (non-cryptographic) XXH128 hash [as discussed here](https://github.com/computational-cell-analytics/micro-sam/issues/36#issuecomment-1809426270)

That's good to know, and we can address this in a follow up.

> * I've removed the 'checkpoint' kwarg from `get_sam_model`. I think it makes more sense to have custom weights of any sort go through the `get_custom_sam_model` function (I haven't touched this function yet), and also if we download weights from different checkpoints then the hash validation is going to break.

I think that makes sense. I need to think about this a bit more because `get_custom_sam_model` currently also supports other kinds of weights. I will read this more carefully later and follow up.

> Problem: I can't test a lot of this, because...
> 
>     * we have low test coverage, and
> 
>     * the example scripts often use hard coded paths to locations on Constantin's computer, so I can't run them (eg: `examples/annotator_with_custom_model.py`)

- Regarding test coverage: we should probably add some more unit-test for the weights, I will think about it.
- Regarding example scripts: good point. I will update those later and remove as much hard-coded paths as possible.

File: micro_sam/util.py
Comment (constantinpape) [inline micro_sam/util.py:180]: I think we should keep this. Merging this into `get_custom_sam_model` would be confusing and I don't see a big issue keeping it here. See my main comment for more details.

File: micro_sam/util.py
Comment (constantinpape) [inline micro_sam/util.py:159]: We can just add something like this to support `checkpoint_path`:
```python
if checkpoint_path is None:
  checkpoint = MODELS.fetch(model_type)
else:
  checkpoint = checkpoint_path
```

Comment (constantinpape) [reply micro_sam/util.py:159]: So I don't see any problem with mixing up hashes.

**Comment by constantinpape** [comment]
Hi @GenevieveBuckley ,
I had some time to look into this now:

## get_sam_model

I think we should keep the (optional) `checkpoint_path` argument there. Mixing this up with `get_custom_sam_model` would be confusing. To clarify this:
- `get_sam_model` loads a model from weights stored in the default format. These can come from two sources: either from weights that are available online (either the original SAM weights or our fine-tuned models) and that we have registered. Or they can be loaded from a **local** filepath that is specified via `checkpoint_path`. 
- `get_custom_sam_model` loads a model from a training checkpoint from our finetuning code. This function is not very relevant for most users, but is important for us to easily evaluate the finetuned models without having to convert them to the default format (which is done by `export_custom_sam_model`)

So these two functions do something conceptually different, and it would also complicate the implementation if `get_custom_sam_model` would accept both types of weight files, because we would then need to find out which of the two formats it is before loading the model.

> * and also if we download weights from different checkpoints then the hash validation is going to break.

We don't want to support downloading weights from arbitrary locations, but only for known models (either original SAM or our finetuned ones). So I don't really see any problems with hash validation here. 
(If a user wants to use some other model available online they can download it themselves and use the `checkpoint_path` argument).

I hope that makes sense to you @GenevieveBuckley.
All of this should probably be documented better, I can follow up with a PR to improve the docstrings after we finish this one.

## test coverage and example scripts

- I have updated the example scripts so that we use `get_cache_directory` in there, see 1c79323
- I have added more tests to extend the coverage for loading models in #280. I will wait for your review before merging since I have written the test for `get_sam_model` according to the current logic.

**Comment by constantinpape** [comment]
Hi @GenevieveBuckley,
I have merged #280 now to increase the test coverage. Most of the changes here should be covered by tests now, so I think you can continue here.
Let me know if something is still unclear with respect to `get_sam_model` vs. `get_custom_sam_model` or if there is some problem in resolving the merge conflicts.

File: micro_sam/util.py
Comment (GenevieveBuckley) [reply micro_sam/util.py]: Done. 
I'm not entirely sure whether the dock widgets get created once when the plugin is imported into napari, or whether a new instance is created every time we open the dock widget from the plugin menu. So I'm not 100% sure whether changes made by the user while napari is open will get picked up by this, but we've done the best we can now.

File: micro_sam/util.py
Comment (GenevieveBuckley) [reply micro_sam/util.py]: Ok. I have restored the `checkpoint_path` keyword argument. 

I will need you to test this (because all the example scripts involving checkpoint_path rely on data you personally have). The `_get_checkpoint` function is not covered by any of the existing tests, so it's safe to say we're not testing anything involving `checkpoint_path` right now. 
Could you please run:
* `examples/annotator_with_custom_model.py`
* `examples/finetuning/use_finetuned_model.py`
* `examples/use_as_library/instance_segmentation.py`

It would be great if we could upload the data and custom weights for `examples/annotator_with_custom_model.py` to zenodo or similar - or even make a similar example but with smaller 2d data and model weights. Ideally then anybody could run this example as a test.

File: micro_sam/util.py
Comment (GenevieveBuckley) [reply micro_sam/util.py]: I'm confused - if the weights are different, the hash value for the object should be different too, right? So how can anybody possibly do some extra fine tuning on a particular model, save the new weights, and then use checkpoint_path and give it the hash value stored in the pooch registry?

I might misunderstand what is actually happening with `checkpoint_path`.

File: micro_sam/util.py
Comment (GenevieveBuckley) [inline micro_sam/util.py:125]: Does `known_hash` need to be None here (i.e. unknown)? I don't quite understand how someone can do fine tuning of model weights, but have the hash stay unchanged (which is I think what would have to happen for what's written here to make any sense).

File: micro_sam/util.py
Comment (GenevieveBuckley) [inline micro_sam/util.py:216]: It seems possible that in the future people might fine tune models based off `vit_t`, so we'd have `model_type="vit_t_em"`, etc. 
I find the `model_type`/`model_type_` distinction here confusing, I'd kinda like to rename `model_type` -> `model_name` instead.

File: micro_sam/util.py
Comment (GenevieveBuckley) [inline micro_sam/util.py:225]: Is this correct?

File: micro_sam/util.py
Comment (constantinpape) [reply micro_sam/util.py]: Hi @GenevieveBuckley,
the models are all available via zenodo already. See https://github.com/computational-cell-analytics/micro-sam/blob/master/micro_sam/util.py#L46-L49. (I agree though that we should list this in the doc, I have added that to #249)
 and I changed the examples in #280 so that all the examples use sample data.

> The `_get_checkpoint` function is not covered by any of the existing tests, so it's safe to say we're not testing anything involving `checkpoint_path` right now.

I don't think that's true either since merging #280. [This](https://github.com/computational-cell-analytics/micro-sam/blob/dev/test/test_util.py#L36-L40) now uses a custom checkpoint. But you would need to merge the current `dev` branch in here first to get those tests.

I will still go ahead and test the examples with this branch.

File: micro_sam/util.py
Comment (constantinpape) [reply micro_sam/util.py]: Sorry I should have written a better description here. I will try to explain what happens if a `checkpoint_path` is passed or not:
- if `checkpoint_path` is given: this is a local path to sam weigths. E.g. if a user has fine-tuned on their own data. In this case we don't download any weights but initialize the model from these local weights. We do not check the weight file against any hashes.
- if `checkpoint_path` is not given: we download the SAM weights corresponding to `model_type`. (I agree that this name is confusing, see comment below). We can check their hash.

> So how can anybody possibly do some extra fine tuning on a particular model, save the new weights, and then use checkpoint_path and give it the hash value stored in the pooch registry?

When finetuning we don't check the weights against the hash values in the pooch registry.

File: micro_sam/util.py
Comment (constantinpape) [reply micro_sam/util.py:135]: Ok, I think I don't fully understand what `pooch` is doing. I will check this out and follow up.

File: micro_sam/util.py
Comment (constantinpape) [reply micro_sam/util.py:242]: I fully agree, we should have a better distinction here and the current distinction is confusing. Let's do what you suggest and rename `model_type -> model_name` and `model_type_ -> model_type`

File: micro_sam/util.py
Comment (constantinpape) [reply micro_sam/util.py:251]: Yes, that's correct.

**Review by constantinpape — APPROVED** [review]

**Comment by constantinpape** [comment]
Thanks @GenevieveBuckley! This cleans up things significantly and I hope now the usage of the `get_sam_model` functions is much clearer.

**Pull request merged**
